serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc

```diff
@@ -89,79 +91,114 @@ The following example shows the declaration of a `java` function:
+
+=== Function Arguments
+
+Your method interface signature should mimic the arguments passed by the flow. 
+
+For example, if you invoke the function using one argument: 
+
+.Example of a `java` function reference with one argument
+[source,json]
+----
+{
+ "functionRef": {
+   "refName": "myFunction",
+   "arguments": {
+     "number": "${.number}"
+   }
+ }
+----
+
+Your method signature should be, assuming `number` model variable is an integer 
+
+.Example of a `java` function implementation
+[source,java]
+----
+public class MyInterfaceOrClass { 
+  
+  public void myMethod(int number) {
+        if (number % 2 != 0) {
+            throw new IllegalArgumentException("Odd situation");
+        }
+    }
+}
+----
+
+As a particular case, if you provide no argument in the workflow definition, the signature of the Java method might include a link:https://github.com/FasterXML/jackson[Jackson's] `JsonNode` parameter. This means that the Java method expects the whole flow model as input.
+
+For example, using this no argument function reference 
+
+.Example of a `java` function reference with no arguments
+[source,json]
+----
+{
+  "functionRef": {
+    "refName": "myFunction"
+   }
+}
+----
+
+If the method signature has a `JsonNode` parameter, as in the following example, the whole model is passed when the method call is performed. 
 
 .Example of a `java` function implementation
 [source,java]
 ----
 public class MyInterfaceOrClass {
 
-    public void myMethod(JsonNode workflowData) {
-        // do whatever I want with the JsonNode:
-        // { "workflowdata": {} }
-    }
 
     public JsonNode myMethod(JsonNode workflowData) {
-        // do whatever I want with the JsonNode:
-        // { "workflowdata": {} }
+        // do whatever I want with the Workflow model
+        ......
         // return the modified content:
         return workflowData;
     }
 }
 ----
 
-To call the function shown in the previous example within your workflow service, you can extract the required JSON value using xref:core/understanding-jq-expressions.adoc[`jq` expression] or pass the JSON value without using arguments.
+=== Function return values
 
-In the following example, the caller sends the entire payload:
+If your method returns a `JsonNode`, the content of that node is merged into the Workflow model (you can use an action data filter to control what is merged)
 
-.Example of a `java` function reference within a `state`
-[source,json]
+The same occurs if your method returns any Java `Object` descendant that is not a primitive wrapper, the Java Object is recursively converted to a JSON one and the result is merged into the Workflow model (you can use an action data filter to control what is merged) 
+
+If your methods returns a primitive type or their corresponding wrapper object (int, boolean, long...), then the primitive value is added to the workflow model with the name `response` (you can change that name using an action data filter) 
```

Comment:
```suggestion
If your method returns a primitive type or their corresponding wrapper object (int, boolean, or long), then the primitive value is added to the workflow model with the name `response` (you can change that name using an action data filter). 
```

Reply:
I cannot commit this, int, boolean and long are three examples of primitive types, but there are more and it makes no sense to enumerate all them here

Reply:
In this case, I'd suggest using "int, boolean, long, and so on" instead of using continuous periods.

Reply:
ok, can you modify and then I will commit suggestion?

Reply:
Done